building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+- A buidling block for cmake
+- Nice non-expert introduction in [py-RSE book, chapter 9](https://merely-useful.tech/py-rse/automate.html)
+- [GNU Make](https://www.gnu.org/software/make/): standard implementation of Make for Linux and macOS
+
+### How does it work?
+
+- When you create / change a file, the OS updates timestamp of file.
+- Make compares timestamps to see which files are older / newer than others.
+- In `Makefile`:
+    - Rules which file(s) depend on which other file(s)
+    - Rules how to update out-of-date files
+- When you run `make`, Make updates out-of-date files including transitive dependencies in right order.
+
+### Single rule example
+
+```makefile
+HelloWorld : main.cpp
+	g++ -o HelloWorld main.cpp
+```
+
+- Explain that `HelloWorld` depends on `main.cpp` and rule to update
+- Three cases:
+    - no `HelloWorld`: run update
+    - `HelloWorld` older than `main.cpp`: run update
+    - `HelloWorld` newer than `main.cpp`: do nothing
+
+- Run make twice
+- `ll` to show timestamp
+
+- Requires actual tabs, spaces not allowed. Show error message.
+
+> Makefile:2: *** missing separator (did you mean TAB instead of 8 spaces?).  Stop.
+
+### Multiple rules example and phony targets
+
+- Add `sse.hpp` and `sse.cpp` in subfolder `sse`
+
+`sse.cpp`:
+
+```cpp
+#include "sse.hpp"
+#include <iostream>
+
+void sse()
+{
+  std::cout << "SSE is great!" << std::endl;
+}
+```
+
+`sse.hpp`:
+
+```cpp
+#pragma once
+
+void sse();
+```
+
+`main.cpp`:
+
+```diff
+#include <iostream>
++ #include "sse.hpp"
+
+int main()
+{
+  std::cout << "Hello World!" << std::endl;
++ sse();
+  return 0;
+}
+```
+
+`Makefile`:
+
+```makefile
+sse.o : sse/sse.hpp sse/sse.cpp
+	g++ -c sse/sse.cpp
+
+HelloWorld : main.cpp sse/sse.hpp
+	g++ -o HelloWorld sse.o main.cpp sse.o
+```
+
+- Run `make`, only builds `sse.o`
+- By default, first target is built
+- `make HelloWorld` to build specific target
+- phony target (doesn't correspond to file, no update rule)
+
+```diff
++ all : HelloWorld sse.o
+```
+
+- Why does Make not just build directly build all targets? We could want to do different things with different targets.
+- Add `clean` target, has no dependency
+
+```diff
++ clean :
++ 	rm HelloWorld sse.o
+```
+
+- Run `make clean`
+- `mkdir clean` and `make clean` confuses Make
+
+```
++ .PHONY : all clean
+```
+
+### Advanced Make
+
+- There is more:
+    - Variables
+    - Rules
+    - Wildcards
+- ... but becomes quickly very hard to read
+- Not covered because cmake does this for us.
+- But nicely documented in [py-RSE chapter 9](https://merely-useful.tech/py-rse/automate.html).
+
+
+## CMake
+
+*35 mins* ... might be tough
+
+### Introduction
+
+
+What is cmake?
+
+> A build system generator, for e.g. makefiles
+
+Why not directly use makefiles?
+
+> CMake can do many more things, is more flexible, avoids complex patterns, and is, thus, more readable
+
+Are there alternatives?
+
+> Yes, `scons`, `autotools`, ...
```

Comment:
meson,  build2 are growing in popularity too.

Reply:
I think here one should be careful what we talk about and what one wants to compare. This kind of mixes two things:

- (Native) Build systems: (GNU) Make, meson, [ninja](https://ninja-build.org/), Visual Studio Files...
  - Software that actually compiles and links source code
- Build system generators: CMake, autotools,...
  - Software that creates files for some build system
  - CMake has generators for [many build systems](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html).

I do not know in what category Scons falls.